jobserver/models/user.py

```diff
@@ -182,6 +133,17 @@ class Meta:
                 name="%(app_label)s_%(class)s_both_pat_expires_at_and_pat_token_set",
             ),
         ]
+        ordering = [
```

Comment:
Have you characterised the performance of this vs the previous ordering? Seems like this is called a lot and we wouldn't want any major performance regressions

Reply:
I've attached some code I've now used to profile this.

This generates many `User` with random `name` and `fullname` from 10-30 characters, with a 5% chance for the `fullname` to be `""` (real server has 500 users and 27 with `fullname` `""`). Then I've tried sorting by each method and recording average time for the sort over several iterations. I've tried 500, 1,000, 10,000 and 100,000 users over 10 iterations to see how they scale from something similar to our current user base to a few orders of magnitude larger, which we'll probably never reach. I've run these a few times each in case of background effects skewing the results. The following runs are representative:

```bash
$ python manage.py profile_user_sort --count 500 --iterations 10 
Average 'order by' ordering time: 0.0082 seconds
Average default ordering time: 0.0038 seconds

$ python manage.py profile_user_sort --count 1000 --iterations 10 
Average 'order by' ordering time: 0.0099 seconds
Average default ordering time: 0.0116 seconds

$ python manage.py profile_user_sort --count 10000 --iterations 10 
Average 'order by' ordering time: 0.0958 seconds
Average default ordering time: 0.1135 seconds

$ python manage.py profile_user_sort --count 100000 --iterations 10 
Average 'order by' ordering time: 0.8913 seconds
Average default ordering time: 0.9592 seconds
```

Conclusion: at 500 users the new default ordering takes about half the time of the old queryset. After doubling in size they're roughly comparable. At much higher user counts the new method scales slightly worse but nothing to be too concerned about.

I suppose the new ordering has slightly lower fixed overheads but slightly worse complexity scaling with length of the list, not sure why.

Reply:
I've attached some code I've now used to profile this.

This generates many `User` with random `name` and `fullname` from 10-30 characters, with a 5% chance for the `fullname` to be `""` (real server has 500 users and 27 with `fullname` `""`). Then I've tried sorting by each method and recording average time for the sort over several iterations. I've tried 500, 1,000, 10,000 and 100,000 users over 10 iterations to see how they scale from something similar to our current user base to a few orders of magnitude larger, which we'll probably never reach. I've run these a few times each in case of background effects skewing the results. The following runs are representative:

```bash
$ python manage.py profile_user_sort --count 500 --iterations 10 
Average 'order by' ordering time: 0.0082 seconds
Average default ordering time: 0.0038 seconds

$ python manage.py profile_user_sort --count 1000 --iterations 10 
Average 'order by' ordering time: 0.0099 seconds
Average default ordering time: 0.0116 seconds

$ python manage.py profile_user_sort --count 10000 --iterations 10 
Average 'order by' ordering time: 0.0958 seconds
Average default ordering time: 0.1135 seconds

$ python manage.py profile_user_sort --count 100000 --iterations 10 
Average 'order by' ordering time: 0.8913 seconds
Average default ordering time: 0.9592 seconds
```

Conclusion: at 500 users the new default ordering takes about half the time of the old queryset. After doubling in size they're roughly comparable. At much higher user counts the new method scales slightly worse but nothing to be too concerned about.

I suppose the new ordering has slightly lower fixed overheads but slightly worse complexity scaling with length of the list, not sure why.

[cmddiff.txt](https://github.com/user-attachments/files/17691236/cmddiff.txt)
[userdiff.txt](https://github.com/user-attachments/files/17691237/userdiff.txt)